internal/pkg/skuba/cni/cilium.go

```diff
@@ -145,8 +254,138 @@ func CreateOrUpdateCiliumConfigMap(client clientset.Interface) error {
 	return nil
 }
 
-func CiliumUpdateConfigMap(client clientset.Interface) error {
-	if err := CreateOrUpdateCiliumConfigMap(client); err != nil {
+// waitForCiliumPod waits until any Cilium preflight pod is running.
+func waitForCiliumPod(client clientset.Interface) (string, error) {
+	var podName string
+
+	klog.Info("waiting for any running Cilium preflight pod")
+
+	// Find any Cilium preflight pod.
+	if err := retry.OnError(migrationRetry, IsErrCiliumNotFound, func() error {
+		pods, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(metav1.ListOptions{
+			LabelSelector: "k8s-app=cilium-pre-flight-check",
+		})
+		if err != nil {
+			return errors.Wrap(err, "could not find cilium preflight pods")
+		}
+		if len(pods.Items) < 1 {
+			return ErrCiliumNotFound
+		}
+		pod := &pods.Items[0]
+		podName = pod.GetName()
+		// Wait until the Cilium pod is not in the pending status and
+		// check whether it's running.
+		klog.Infof("waiting for availability of cilium preflight pod %s", podName)
+		for {
+			var err error
+
+			if pod.Status.Phase != v1.PodPending {
+				break
+			}
+
+			time.Sleep(15 * time.Second)
+			pod, err = client.CoreV1().Pods(metav1.NamespaceSystem).Get(podName, metav1.GetOptions{})
+			if err != nil {
+				return errors.Wrapf(ErrCiliumNotFound, "could not get cilium preflight pod: %v", err)
+			}
+		}
```

Comment:
A problem I had in my env is that the check was too short. The preflight pods were coming later but 15 seconds was too small. I changed this for to:
```
                var myerr error
                for i := 1; i <= 10; i++ {
                        if pod.Status.Phase != v1.PodPending {
                                break
                        }

                        time.Sleep(15 * time.Second)
                        pod, err = client.CoreV1().Pods(metav1.NamespaceSystem).Get(podName, metav1.GetOptions{})
                        if myerr != nil {
                                fmt.Println("Preflight pods not ready yet! Trying again")
                        }
                }
                if myerr != nil {
                        return errors.Wrapf(ErrCiliumNotFound, "could not get cilium preflight pod: %v", err)
                }
```
And then it worked!
```
I0421 17:41:30.201149   21638 cilium.go:314] starting migration from etcd to CRD as a data store for cilium
I0421 17:41:30.201248   21638 cilium.go:261] waiting for any running Cilium preflight pod
I0421 17:41:30.805319   21638 cilium.go:278] waiting for availability of cilium preflight pod cilium-pre-flight-check-6h9pl
I0421 17:41:30.805353   21638 cilium.go:302] found Cilium preflight pod
I0421 17:41:31.882737   21638 cilium.go:357] successfully migrated from etcd to CRD
```

Reply:
The problem in your snippet is that nothing gets assigned to `myerr` variable - it will be always `nil`. But the forloop might help.

Reply:
Also, please notice the `retry.OnError` utility. It should retry on each `ErrCiliumNotFound`multiple times. But maybe it proves that `retry.OnError` doesn't work really well and maybe we should have our own retry logic...

Reply:
OK, now I realized that `k8s.io/client-go/util/retry` retries only **twice** by default, so it's useless for us. I will look for a better library or write my own logic.